### PR TITLE
Suggest removing `&`s

### PR DESCRIPTION
**Pull request by ysiraichi**
This implements the error message discussed in #47744.
We check whether removing each `&` yields a type that satisfies the requested obligation.
Also, it was created a new `NodeId` field in `ObligationCause` in order to iterate through the `&`s. The way it's implemented now, it iterates through the obligation snippet and counts the number of `&`.

r? @estebank

### CODE REVIEW

File: src/librustc/traits/error_reporting.rs
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]:
```
} else if !c.is_whitespace() {
    break;
}
```
to avoid iterating over the entire snippet (which can be arbitrarily large) and to account for code like `& & & x`.

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: This can be written as

```
let refs_number = snippet.chars().take_while(|c| c == '&' || !c.is_whitespace()).filter(|c| c == '&').count();
```

File: src/librustc/traits/error_reporting.rs
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: wrong indentation

File: src/librustc/traits/mod.rs
Comment (estebank) [inline src/librustc/traits/mod.rs]: I haven't found any call to `new` that didn't use the same thing for `body_id` and `node_id`. Did I miss something?

Comment (ysiraichi) [reply src/librustc/traits/mod.rs]: For now, I changed the code so that I could later fill `node_id` with the appropriate node (so that I wouldn't have to iterate the snippet).

File: src/librustc/traits/error_reporting.rs
Comment (ysiraichi) [reply src/librustc/traits/error_reporting.rs]: Shouldn't it be:
```
let refs_number = snippet.chars().filter(|c| !c.is_whitespace()).take_while(|c| c == '&').count();
```

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: I made a mistake, you want either & or whitespace, my code gives you only &. Your code would  indeed work.

File: src/librustc/traits/error_reporting.rs
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: This doesn't account for possible whitespace between the `&`: `& &    & x`, for example.

File: src/librustc/traits/error_reporting.rs
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: This is a bit more vague than I'd like, can we teak the sentence depending on the amount of `&`s that should be removed?

File: src/librustc/traits/error_reporting.rs
Comment (ysiraichi) [reply src/librustc/traits/error_reporting.rs]: Oh. You are right! I will fix it.

File: src/librustc/traits/error_reporting.rs
Comment (ysiraichi) [reply src/librustc/traits/error_reporting.rs]: How does "consider removing N references `&`", where N is the number of references to be removed?

File: src/librustc/traits/error_reporting.rs
Comment (estebank) [inline src/librustc/traits/error_reporting.rs]: Wouldn't this turn something like `& ( a, b )` into `(a,b)` or `& & & & v   .iter()    .enumerate()` into `v.iter().enumerate()`?

File: src/test/ui/suggest-remove-refs-2.rs
Comment (estebank) [inline src/test/ui/suggest-remove-refs-2.rs]: Could you also add another test where the above uses multiple lines?

```rust
for (i, n) in & & & &v
    .iter()
    .enumerate()
{
    println!("{} {}", i, n);
}
```

**Review by estebank — COMMENTED** [review]
Added some comments to handle some corner cases. Like it a lot and can't wait to see it merged!

File: src/librustc/traits/error_reporting.rs
Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: So, for things like tuples, arrays and slices, we would like to keep the current formatting. For things like very long chains of method calls split each on its own line (because having them all in the same line would be over 100 chars wide) we would also like to keep the formatting.

I think you should instead advance the iterator while the char is whitespace or `&`, and return the everything after that. That way you trim the beginning the way you want, but keep the rest of the code with the current formatting.

Comment (ysiraichi) [reply src/librustc/traits/error_reporting.rs]: Agreed.

File: src/test/ui/suggest-remove-refs-3.stderr
Comment (estebank) [inline src/test/ui/suggest-remove-refs-3.stderr]: 😍

File: src/test/ui/suggest-remove-refs-2.stderr
Comment (estebank) [inline src/test/ui/suggest-remove-refs-2.stderr]: After travis passes, I'm happy to merge, but I think we could tweak this wording slightly to make it clearer.

@nikomatsakis, @Mark-Simulacrum are you ok with this wording, or would you prefer something like

```
consider removing 5 `&`-references
```

or

```
consider removing the 5 leading `&`
```

Comment (Mark-Simulacrum) [reply src/test/ui/suggest-remove-refs-2.stderr]: I think we definitely want the "references" in there, and I think we should remove `v.iter().enumerate()` from the help since it somewhat looks odd (since it's not what we're removing). Otherwise though I think this is good as-is.

Comment (ysiraichi) [reply src/test/ui/suggest-remove-refs-2.stderr]: What about using `node_id` from `ObligationCause`? I am still searching only for occurrences of `&` in the snippet.
In case we don't need to use `node_id`, I should remove it from the codebase. Don't you think?

Comment (estebank) [reply src/test/ui/suggest-remove-refs-2.stderr]: @Mark-Simulacrum we could replace the `span_suggestion` with `span_suggestion_short` so that the code is not being shown in the label when displayed inline, the suggestion is displayed in full when it is shown on its own (like the multiline case above) and the tooling will be able to apply the correct output as it would now.

@ysiraichi would you mind doing that change? Once that is done, r=me.

Comment (ysiraichi) [reply src/test/ui/suggest-remove-refs-2.stderr]: Sure... Which one?

Comment (estebank) [reply src/test/ui/suggest-remove-refs-2.stderr]: `span_suggestion` -> `span_suggestion_short` and

```
consider removing the 5 leading `&`-references
```

This line would end up looking like this:

```
LL |     for (i, n) in & & & & &v.iter().enumerate() {
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                   |
   |                   `&&&&&std::iter::Enumerate<std::slice::Iter<'_, {integer}>>` is not an iterator; maybe try calling `.iter()` or a similar method
   |                   help: consider removing the 5 leading `&`-references
```

If you had the extra time and felt it was worth it, it'd be great to change the suggestion's span to be

```
LL |     for (i, n) in & & & & &v.iter().enumerate() {
   |                   ---------^^^^^^^^^^^^^^^^^^^^^
   |                   |
   |                   `&&&&&std::iter::Enumerate<std::slice::Iter<'_, {integer}>>` is not an iterator; maybe try calling `.iter()` or a similar method
   |                   help: consider removing the 5 leading `&`-references
```

so that the suggestion will have a different color. That being said, I _won't_ block this PR on that.

File: src/libsyntax/codemap.rs
Comment (ysiraichi) [inline src/libsyntax/codemap.rs]: Not sure if I should have done this here or inlined it in the function `suggest_remove_refs`. It seemed right to do it here... I am considering two options now:
- inline this function into the only place it's being used; or
- use this function in functions like `span_until_whitespace` which may use it.

@estebank what do you think?

Comment (estebank) [reply src/libsyntax/codemap.rs]: It's good to separate it, it makes the code above easier to follow.

The latter option sounds good to me if you don't mind doing it :)

File: src/test/ui/suggest-remove-refs-2.stderr
Comment (estebank) [inline src/test/ui/suggest-remove-refs-2.stderr]: The output is great!

**Review by estebank — APPROVED** [review]

File: src/librustc/traits/mod.rs
Comment (ysiraichi) [reply src/librustc/traits/mod.rs]: By the way, are we going to leave these changes (adding `node_id` to `ObligationCause`) even if we are not really using it (yet)?

Comment (estebank) [reply src/librustc/traits/mod.rs]: Given it'd be dead code, lets remove it and land this PR. We can make that change again once the `node_id` is actually used.

File: src/test/ui/suggest-remove-refs-2.stderr
Comment (nikomatsakis) [reply src/test/ui/suggest-remove-refs-2.stderr]: yeah, this is good

**Pull request merged**
